besu/src/test/java/org/hyperledger/besu/PrivacyReorgTest.java

```diff
@@ -0,0 +1,558 @@
+            .targetGasLimit(GasLimitCalculator.DEFAULT)
+            .build();
+  }
+
+  @Test
+  public void privacyGroupHeadIsTracked() {
+    // Setup an initial blockchain with one private transaction
+    final ProtocolContext<?> protocolContext = besuController.getProtocolContext();
+    final DefaultBlockchain blockchain = (DefaultBlockchain) protocolContext.getBlockchain();
+    final PrivateStateStorage privateStateStorage = privacyParameters.getPrivateStateStorage();
+
+    final Block firstBlock =
+        gen.block(
+            new BlockDataGenerator.BlockOptions()
+                .setBlockNumber(1)
+                .setParentHash(blockchain.getGenesisBlock().getHash())
+                .addTransaction(generateMarker(getEnclaveKey(enclave.clientUrl())))
+                .hasOmmers(false)
+                .setReceiptsRoot(
+                    Hash.fromHexString(
+                        "0xc8267b3f9ed36df3ff8adb51a6d030716f23eeb50270e7fce8d9822ffa7f0461"))
+                .setGasUsed(23176)
+                .setLogsBloom(LogsBloomFilter.empty())
+                .setStateRoot(
+                    Hash.fromHexString(
+                        "0x16979b290f429e06d86a43584c7d8689d4292ade9a602e5c78e2867c6ebd904e")));
+
+    appendBlock(besuController, blockchain, protocolContext, firstBlock);
+
+    final PrivacyGroupHeadBlockMap expected =
+        new PrivacyGroupHeadBlockMap(
+            Collections.singletonMap(
+                Bytes32.fromHexString(
+                    "0xf250d523ae9164722b06ca25cfa2a7f3c45df96b09e215236f886c876f715bfa"),
+                firstBlock.getHash()));
+
+    assertThat(
+            privateStateStorage.getPrivacyGroupHeadBlockMap(blockchain.getGenesisBlock().getHash()))
+        .isEmpty();
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash())).isNotEmpty();
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash()))
+        .contains(expected);
+
+    final Block secondBlock =
+        gen.block(
+            new BlockDataGenerator.BlockOptions()
+                .setBlockNumber(2)
+                .setParentHash(firstBlock.getHeader().getHash())
+                .hasTransactions(false)
+                .hasOmmers(false)
+                .setReceiptsRoot(
+                    Hash.fromHexString(
+                        "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421"))
+                .setGasUsed(0)
+                .setLogsBloom(LogsBloomFilter.empty())
+                .setStateRoot(
+                    Hash.fromHexString(
+                        "0xd86a520e49caf215e7e4028262924db50540a5b26e415ab7c944e46a0c01d704")));
+
+    appendBlock(besuController, blockchain, protocolContext, secondBlock);
+
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(secondBlock.getHash())).isNotEmpty();
+
+    assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(firstBlock.getHash()))
```

Comment:
I'm having a hard time trying to understand this test. If we want to test that the privateStateStorage won't change because the second block doensn't have a private transaction, we should probably change this to:
```
assertThat(privateStateStorage.getPrivacyGroupHeadBlockMap(secondBlock.getHash())).contains(expected);
```

Reply:
yes you are right